src/analysis/normed/ring/seminorm_constructions.lean

```diff
@@ -0,0 +1,328 @@
+/-
+Copyright (c) 2022 María Inés de Frutos-Fernández. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: María Inés de Frutos-Fernández
+-/
+import analysis.special_functions.pow
+import analysis.normed.ring.seminorm
+import topology.metric_space.basic
+import topology.algebra.order.monotone_convergence
+
+/-!
+# Smoothing procedures for seminorms
+
+This file contains several ring seminorm constructions that will be used in the proof that a rank 1
+nonarchimedean valuation on a complete field can be uniquely extended to any algebraic extension.
+
+## Main declarations
+
+* `seminorm_from_const`: Given a power-multiplicative ring seminorm `f` on a ring `R` and an
+  element `c ∈ R` with `f c ≠ 0`, the real-valued function sending `x ∈ R` to the limit of
+  `(f (x * c^n))/((f c)^n)` is a power-multiplicative ring seminorm on `R` satisfying some
+  multiplicativity conditions [BGR84, Proposition 1.3.2/2].
+
+
+## References
+
+* [S. Bosch, U. Güntzer, R. Remmert, *Non-Archimedean Analysis*][bosch-guntzer-remmert]
+
+## Tags
+ring_seminorm, ring_norm
+-/
+
+
+noncomputable theory
+
+open_locale topological_space
+
+section defs
+
+/-- A function `f : α → β` is power-multiplicative if for all `a ∈ α` and all positive `n ∈ ℕ`,
+  `f (a ^ n) = (f a) ^ n`. -/
+def is_pow_mul {α β : Type*} [has_pow α ℕ] [has_pow β ℕ]  (f : α → β) :=
+∀ (a : α) {n : ℕ} (hn : n ≠ 0), f (a ^ n) = (f a) ^ n
+
+/-- A function `f : α → β` is nonarchimedean if it satisfies the inequality
+  `f (a + b) ≤ max (f a) (f b)` for all `a, b ∈ α`. -/
+def is_nonarchimedean {α : Type*} [has_add α] {β : Type*} [linear_order β] (f : α → β) : Prop :=
+∀ r s, f (r + s) ≤ max (f r) (f s)
+
+end defs
+
+section seminorm_from_const
+/- In this section we prove Proposition 1.3.2/2 in
+[S. Bosch, U. Güntzer, R. Remmert, *Non-Archimedean Analysis*][bosch-guntzer-remmert]. -/
+
+section ring
+
+variables {R : Type*} [comm_ring R] (c : R) (f : ring_seminorm R) (hc : 0 ≠ f c)
+  (hpm : is_pow_mul f)
+
+/-- For a ring seminorm `f` on `R` and `c ∈ R`, the sequence given by `(f (x * c^n))/((f c)^n)`. -/
+def seminorm_from_const_seq (x : R) : ℕ → ℝ :=
```

Comment:
I think it is much more readable to include all explicit variables here, especially since casual readers could think `hc` will become an argument of this function.

Reply:
@sgouezel once told me that using `variables` for `_ : Prop` assumptions is usually a bad idea.